.github/workflows/ci.yml

```diff
@@ -41,25 +41,36 @@ jobs:
     - uses: ./
 
     - name: Install dependencies
-      run: sudo apt-get update && sudo apt-get install --no-install-recommends python3-pexpect
+      run: sudo apt-get update && sudo apt-get install --no-install-recommends python3-pexpect libnss-mymachines libnss-resolve libnss-systemd
 
     # Do a manual install so we have the latest changes from the pull request available.
     - name: Install
       run: sudo python3 -m pip install .
 
-    - name: Build ${{ matrix.distro }}/${{ matrix.format }}
+    - name: Write shared mkosi config
       run: |
-        tee mkosi.default << EOF
+        mkdir -p mkosi.default.d
+
+        sudo systemctl enable --now systemd-networkd
```

Comment:
Why is it necessary to enable it on the build host?

Reply:
I guess this is for the `resolvectl query` in the `tests/pexpect/boot.py`.

Reply:
`mkosi ssh` works by sshing into the hostname we assigned to the container. Of course, that hostname has to be resolved first to the IP address of the container/vm. To make that work, the container/vm needs to have an IP address. This is why we enable `NetworkVeth` so that we have a Veth/Tap link to the container.

A Veth/Tap link isn't sufficient though to give the container/vm an IP address, we also need a DHCP server to hand out an IP. This is why we enable systemd-networkd. networkd has a default network file that matches TAP devices on the host with a specific prefix and starts a DHCP server on them to hand out IP addresses (see https://github.com/systemd/systemd/blob/master/network/80-vm-vt.network). mkosi makes sure the TAP device is named so that systemd-networkd automatically starts a DHCP server on the host side of the TAP device. 

Once the container/vm has an IP address, we also need that IP address to be associated with the container/vm's hostname. When using systemd-nspawn, this happens automatically as systemd-nspawn registers the container along with its hostname and the ifindex of its veth link with systemd-machined which uses this information to allow resolving the hostname to the link local address of the container via nss-mymachines.

Theoretically, this should be possible with QEMU as well, but QEMU doesn't integrate with systemd-machined yet. I actually started on this but dropped it since I found a quicker solution. That is, once systemd-resolved is running in the container, it advertises the hostname of the container/vm via LLMNR by default. If systemd-resolved is running on the host, it picks this up and the hostname of the container/vm can be resolved to its IP address.

Does that make sense? networkd runs the DHCP server, resolved does LLMNR to make sure we can resolve the hostname of the container/vm on the host.

Reply:
That makes sense, although requiring changes on the host is a bit too much IMHO. I am not going to override the default the distro sets for my machine, that just spells trouble. On Debian we have network-manager for desktops and ifupdown for headless installations.

What about the qemu config for usermode emulated networking I suggested here:

https://github.com/systemd/mkosi/issues/593#issuecomment-750867664

With that, you don't need to do anything on the host. In fact I've used that for my lightweight VMs (ie: no libvirt) for ages now.

Reply:
networkd should be non-intrusive by default. In other works, it won't try to manage any network interfaces except those it is explicitly told to manage and a few network interfaces that follow some specific naming convention. Of course, if the other network manager you've got running tries to manage all interfaces that appear on the host, there will be trouble as both networkd and the other network manager will be trying to manage the same network interface.

Although really, we shouldn't care too much about what's running on the host, as long as in the end, the hostname of the container/vm is resolvable. We might not even need networkd on the host if LLMNR advertises link local addresses. We just need whatever network manager on the host to pick up LLMNR and make it available via NSS. If LLMNR doesn't work, we can also configure resolved in the vm/container to do MulticastDNS as well.

I can also get back to integrating QEMU with systemd-machined but that does require nss-mymachines to be configured in nssswitch on the host.

Do you know if Debian makes hostnames received via LLMNR/mDNS available via NSS? I know resolved does it but I'm not sure if what Debian uses does it as well.

Reply:
Reworked to use port forwarding as that doesn't require as much of the host. I started with the hostname solution because hostnames are easier to remember than port names but since we have `mkosi ssh` regardless, it's not a big problem anymore to use hostnames instead of ports since `mkosi ssh` hides that for us.

Reply:
> As for the user mode solution, I'd prefer solutions that work the same across containers/vms and play well with NetworkVeth. 

Up to you of course, but personally I wouldn't mind if the implementation uses different methods. After all, qemu and nspawn are different things. As long as the user interface in mkosi is the same. I don't see an issue there.

I quickly tried your branch and it looks nice! A couple of things:

- for Debian we still need to enable networkd/resolved in the image, and drop in a .network file that matches the interface
- when I tried mkosi ssh, it could connect but then it asked for a passphrase for the ssh key:
```
$ sudo bin/mkosi --default mkosi.files/mkosi.debian ssh
[sudo] password for luca: 
Enter passphrase for key '/home/luca/git/mkosi/mkosi.output/id_rsa':
```

What's the passphrase set to?

Reply:
> What's the passphrase set to?

`args.password` or `""` if that's empty

Reply:
Ah right, thanks - that works!

Reply:
I'm running blank on a good way to support this for nspawn. Using hostnames requires the host to support LLMNR, Using port forwarding doesn't work with nspawn. Using machinectl shell requires users to enter their password (which wouldn't be required or would be a different password than the on required for SSH with qemu). Fixing nspawn will take ages to become available in all distros. 

Maybe we can hack in our own iptables rule to make port forwarding with nspawn work with localhost?

Reply:
I'm not sure it makes sense to bend over backwards. I think it's fine to document under what circumstances this will work and how these can be achieved, i.e. requiring networkd. That should be less surprising than setting iptable rules.